rpc/datapool.go

```diff
@@ -112,27 +119,140 @@ func (p *DataPool) GetCacheOrResolveBNS(deviceName string, client *Client) ([]Ad
+				p.bnsCacheUpdatingFlag[peerKey] = true
+				go p.updateCacheResolvePeers(deviceName, client)
+			}
+		}
+		return peers, nil
+	}
+
+	p.updateCacheResolvePeers(deviceName, client)
+
+	peers, cached = p.getCacheBNS(peerKey)
+	if cached {
+		return peers, nil
+	}
+
+	return nil, nil
+
+}
+
+// getCacheOrResolvePeersOfDrive get peers of drive address
+func (p *DataPool) GetCacheOrResolvePeersOfDrive(addr Address, client *Client) ([]Address, error) {
+
+	peerKey := fmt.Sprintf("peers:%s", addr.HexString())
+	peers, cached := p.getCacheBNS(peerKey)
+	if cached {
+		//check if cache is expired if so call updateCacheResolvePeersOfDrive async. Expire duration is config.AppConfig.ResolveCacheTime
+		if !p.bnsCacheUpdatingFlag[peerKey] && config.AppConfig.ResolveCacheTime > 0 {
+			if expireTime, ok := p.bnsCacheExpireItem[peerKey]; ok && time.Now().After(expireTime) {
+				p.bnsCacheUpdatingFlag[peerKey] = true
+				go p.updateCacheResolvePeersOfDrive(addr, client)
+			}
+		}
+		return peers, nil
+	}
+
+	p.updateCacheResolvePeersOfDrive(addr, client)
+
+	peers, cached = p.getCacheBNS(peerKey)
+	if cached {
+		return peers, nil
+	}
+
+	return nil, nil
+}
+
+// getCacheOrResolveMembersOfDriveMember get members of drive member address
+func (p *DataPool) GetCacheOrResolveMembersOfDriveMember(addr Address, client *Client) ([]Address, error) {
+	peerKey := fmt.Sprintf("members:%s", addr.HexString())
+	peers, cached := p.getCacheBNS(peerKey)
+	if cached {
+		//check if cache is expired if so call updateCacheResolveMembersOfDriveMember async. Expire duration is config.AppConfig.ResolveCacheTime
+		if !p.bnsCacheUpdatingFlag[peerKey] && config.AppConfig.ResolveCacheTime > 0 {
+			if expireTime, ok := p.bnsCacheExpireItem[peerKey]; ok && time.Now().After(expireTime) {
+				p.bnsCacheUpdatingFlag[peerKey] = true
+				go p.updateCacheResolveMembersOfDriveMember(addr, client)
+			}
+		}
+		return peers, nil
+	}
+
+	p.updateCacheResolveMembersOfDriveMember(addr, client)
+
+	peers, cached = p.getCacheBNS(peerKey)
 	if cached {
 		return peers, nil
 	}
 
+	return nil, nil
+}
+
+func (p *DataPool) updateCacheResolveMembersOfDriveMember(addr Address, client *Client) {
+	peerKey := fmt.Sprintf("members:%s", addr.HexString())
+	var peersList []Address
+	members, err := client.ResolveMembers(addr)
+	if err == nil {
+		peersList = append(peersList, members...)
+	}
+
+	p.SetCacheBNS(peerKey, peersList)
+	p.bnsCacheExpireItem[peerKey] = time.Now().Add(config.AppConfig.ResolveCacheTime)
+	p.bnsCacheUpdatingFlag[peerKey] = false
+}
+
+func (p *DataPool) updateCacheResolvePeersOfDrive(addr Address, client *Client) {
+	peerKey := fmt.Sprintf("peers:%s", addr.HexString())
+	var peersList []Address
+	members, err := client.ResolveMembers(addr)
+	if err == nil {
+		peersList = append(peersList, members...)
+	}
+	for _, address := range members {
+		members, new_err := client.ResolveMembers(address)
```

Comment:
In regards to the previous comment I think you can change this function to work recursively for both drives and members by making this call dependent on failure / success like you have in the `ResolveAccountType`

```go
func (p *DataPool) resolveAllMembers(members Address[], client *Client) ([]Address) {
	for _, maybeMemberAddr := range members {
			members, new_err := client.ResolveMembers(maybeMemberAddr)
			if new_err == nil {
	                        // smart contract, might need to recurse
				peersList = append(peersList, resolveAllMembers(members)...)
			} else {
	                        // not a smart contract, instead a real (device) peer
				peersList = append(peersList, maybeMemberAddr) 
	                }
		}
	}
}
```